airbyte-integrations/connectors/source-doubleverify/source_doubleverify/source.py

```diff
@@ -0,0 +1,284 @@
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+import logging
+import time
+
+
+# Basic full refresh stream
+class DoubleverifyStream(HttpStream, ABC):
+
+    url_base = "https://data-api.doubleverify.com"
+    request_type = None
+    logger = logging.getLogger('airbyte')
+
+    def __init__(self, config: Mapping[str, Any], catalog_stream: Mapping[str, Any]):
+        super().__init__()
+        self.config = config
+        self.catalog_stream = catalog_stream
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        """
+        Override to return any non-auth headers. Authentication headers will overwrite any overlapping headers returned from this method.
+        """
+        return Utils.get_request_header(config=self.config, accept='text/csv', accept_encoding='gzip')
+        
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        This method should return a Mapping (e.g: dict) containing whatever information required to make paginated requests. This dict is passed
+        to most other methods in this class to help you form headers, request bodies, query params, etc..
+
+        For example, if the API accepts a 'page' parameter to determine which page of the result to return, and a response from the API contains a
+        'page' number, then this method should probably return a dict {'page': response.json()['page'] + 1} to increment the page count by 1.
+        The request_params method should then read the input next_page_token and set the 'page' param to next_page_token['page'].
+
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def path(self, **kwargs) -> str:
+        request_id = self.post_create_request(self.config, self.catalog_stream)
+        #check if the report is successfully created
+        if request_id[0]:
+            # Get the report status
+            report_status = self.poll_report_status(request_id[1], self.config)
+            # If the report status is success, then return the correct URL
+            if report_status == "Success":
+                # Introduce a small delay to wait for the csv report to be generated on the server side.
+                time.sleep(5)
+                return DoubleverifyStream.url_base + "/requests/" + request_id[1] + "/data"
```

Comment:
What happen if you don't have the index? This should be some iterative method?

Reply:
The method `post_create_request()` returns a `Tuple[Boolean, Any]`, so we will always have the index. 
If the report creation was successful, `request_id[0]` would be True, and `request_id[1]` would have the id of the created report. Otherwise,  `request_id[0]` would be False, and `request_id[1] `would return either the error message of the failure reason.